File: python/telequick_agents/wire.py
"""TeleQuick agent-transport wire protocol.

Two lanes over one WebTransport/QUIC session (frozen spec):

  * MEDIA — WebTransport **datagrams**. Fixed 6-byte header + payload:

        version(1) type(1) call_id(2, BE) seq(2, BE)  | payload

    `call_id` demuxes concurrent calls on a shared session; `seq` lets the
    receiver detect loss/reorder (a dropped datagram is a lost 20 ms, never a
    stall). The audio codec + sample rate are negotiated ONCE on the control
    stream (job_assign.audio), so datagrams stay lean and carry only payload.

  * CONTROL — a bidi **stream** of newline-delimited JSON envelopes, same shape
    style as the Pusher frames mod_realtime already speaks:

        {"op":"job_assign", "call_id":.., "room":.., "audio":{...}, ...}
        {"op":"flush","call_id":..}  {"op":"clear","call_id":..}
        {"op":"playout","call_id":..,"position":..,"interrupted":..}
        {"op":"shutdown","call_id":..}   {"op":"register",...} {"op":"ready"}

This module is pure (no I/O) so the framing is unit-tested without a socket;
`transport.WebTransportSession` binds it to an aioquic connection.
"""

from __future__ import annotations

import json
from dataclasses import dataclass

WIRE_VERSION = 1

# datagram types
DG_AUDIO = 1

_HEADER_LEN = 6


@dataclass
class Datagram:
    call_id: int
    seq: int
    payload: bytes
    type: int = DG_AUDIO
    version: int = WIRE_VERSION


def encode_datagram(dg: Datagram) -> bytes:
    if not (0 <= dg.call_id <= 0xFFFF):
        raise ValueError(f"call_id out of range: {dg.call_id}")
    if not (0 <= dg.seq <= 0xFFFF):
        raise ValueError(f"seq out of range: {dg.seq}")
    header = bytes(
        (
            dg.version & 0xFF,
            dg.type & 0xFF,
            (dg.call_id >> 8) & 0xFF,
            dg.call_id & 0xFF,
            (dg.seq >> 8) & 0xFF,
            dg.seq & 0xFF,
        )
    )
    return header + dg.payload


def decode_datagram(buf: bytes) -> Datagram:
    if len(buf) < _HEADER_LEN:
        raise ValueError(f"short datagram: {len(buf)} bytes")
    version, type_ = buf[0], buf[1]
    if version != WIRE_VERSION:
        raise ValueError(f"unsupported wire version {version}")
    call_id = (buf[2] << 8) | buf[3]
    seq = (buf[4] << 8) | buf[5]
    return Datagram(call_id=call_id, seq=seq, payload=buf[_HEADER_LEN:], type=type_, version=version)


def encode_ctrl(op: str, **fields) -> bytes:
    """One newline-terminated control envelope."""
    obj = {"op": op, **fields}
    return (json.dumps(obj, separators=(",", ":")) + "\n").encode()


def decode_ctrl_stream(buf: bytes) -> tuple[list[dict], bytes]:
    """Parse whole envelopes from a byte buffer.

    Returns (messages, remainder) — the remainder is a partial trailing line to
    be prepended to the next read. Robust to envelopes split across QUIC STREAM
    frames.
    """
    msgs: list[dict] = []
    while (nl := buf.find(b"\n")) != -1:
        line, buf = buf[:nl], buf[nl + 1 :]
        line = line.strip()
        if line:
            msgs.append(json.loads(line))
    return msgs, buf

File: wire.go
// Package agents is the TeleQuick external-agent SDK for Go: run your own
// voice agent process against real phone calls over the TeleQuick QUIC media
// plane. One outbound QUIC/WebTransport session (no inbound ports), HMAC auth
// with your media app key/secret, one Call per routed phone call.
//
// It is the Go sibling of the Python `telequick-agents` and TypeScript
// `@telequick/agents` packages — same wire protocol, same handshake, same
// Serve(handler) shape.
package agents

import (
	"bytes"
	"encoding/json"
	"errors"
	"fmt"
)

// Wire protocol — byte-identical to the Python/TS packages and the engine.
//
//	MEDIA   — WebTransport datagrams: version(1) type(1) call_id(2,BE) seq(2,BE) | payload
//	CONTROL — one bidi stream of newline-delimited JSON envelopes {"op":...,...}
const (
	WireVersion byte = 1
	DgAudio     byte = 1
	headerLen        = 6
)

// Datagram is one media frame on the wire.
type Datagram struct {
	CallID  uint16
	Seq     uint16
	Payload []byte
	Type    byte
	Version byte
}

// MakeDatagram builds an audio datagram for callID/seq.
func MakeDatagram(callID, seq uint16, payload []byte) Datagram {
	return Datagram{CallID: callID, Seq: seq, Payload: payload, Type: DgAudio, Version: WireVersion}
}

// EncodeDatagram serializes the 6-byte header + payload.
func EncodeDatagram(dg Datagram) []byte {
	out := make([]byte, headerLen+len(dg.Payload))
	out[0] = dg.Version
	out[1] = dg.Type
	out[2] = byte(dg.CallID >> 8)
	out[3] = byte(dg.CallID)
	out[4] = byte(dg.Seq >> 8)
	out[5] = byte(dg.Seq)
	copy(out[headerLen:], dg.Payload)
	return out
}

// DecodeDatagram parses a datagram; the payload is a copy.
func DecodeDatagram(buf []byte) (Datagram, error) {
	if len(buf) < headerLen {
		return Datagram{}, fmt.Errorf("short datagram: %d bytes", len(buf))
	}
	if buf[0] != WireVersion {
		return Datagram{}, fmt.Errorf("unsupported wire version %d", buf[0])
	}
	payload := make([]byte, len(buf)-headerLen)
	copy(payload, buf[headerLen:])
	return Datagram{
		Version: buf[0],
		Type:    buf[1],
		CallID:  uint16(buf[2])<<8 | uint16(buf[3]),
		Seq:     uint16(buf[4])<<8 | uint16(buf[5]),
		Payload: payload,
	}, nil
}

// CtrlMsg is one decoded control envelope.
type CtrlMsg map[string]any

// Op returns the envelope's "op", or "" when absent.
func (m CtrlMsg) Op() string {
	s, _ := m["op"].(string)
	return s
}

// EncodeCtrl builds one newline-terminated control envelope. "op" is
// serialized first (matches the Python/TS encoders); the rest of the fields
// follow in encoding/json's key order.
func EncodeCtrl(op string, fields map[string]any) ([]byte, error) {
	var buf bytes.Buffer
	buf.WriteString(`{"op":`)
	opb, err := json.Marshal(op)
	if err != nil {
		return nil, err
	}
	buf.Write(opb)
	if len(fields) > 0 {
		fb, err := json.Marshal(fields)
		if err != nil {
			return nil, err
		}
		buf.WriteByte(',')
		buf.Write(fb[1:]) // drop the leading '{' — fb already ends with '}'
	} else {
		buf.WriteByte('}')
	}
	buf.WriteByte('\n')
	return buf.Bytes(), nil
}

// DecodeCtrlStream parses whole envelopes from a growing byte buffer and
// returns them plus the unconsumed remainder (a partial trailing line to be
// prepended to the next read). Robust to envelopes split across QUIC STREAM
// frames. A malformed line is skipped, not fatal.
func DecodeCtrlStream(buf []byte) ([]CtrlMsg, []byte) {
	var msgs []CtrlMsg
	for {
		nl := bytes.IndexByte(buf, '\n')
		if nl < 0 {
			break
		}
		line := bytes.TrimSpace(buf[:nl])
		buf = buf[nl+1:]
		if len(line) == 0 {
			continue
		}
		var m CtrlMsg
		if err := json.Unmarshal(line, &m); err == nil {
			msgs = append(msgs, m)
		}
	}
	rem := make([]byte, len(buf))
	copy(rem, buf)
	return msgs, rem
}

var errShort = errors.New("short read")
